orbit/cmd/orbit/orbit.go

```diff
@@ -939,44 +936,81 @@ func (d *desktopRunner) interrupt(err error) {
 	}
 }
 
-// shell out to osquery (on Linux and macOS) or to wmic (on Windows), and get
-// the system uuid and serial number
-func getUUIDAndSerial(osqueryPath string) (uuid, serial string, err error) {
+// hostInfo is used to parse osquery JSON output from `system_info` and `os_version` tables.
+type hostInfo struct {
+	// HardwareUUID is the unique identifier for this device (extracted from `system_info` osquery table).
+	HardwareUUID string `json:"uuid"`
+	// HardwareSerial is the unique serial number for this device (extracted from `system_info` osquery table).
+	HardwareSerial string `json:"hardware_serial"`
+	// Hostname is the device's hostname (extracted from `system_info` osquery table).
+	Hostname string `json:"hostname"`
+	// Platform is the device's platform as defined by osquery (extracted from `os_version` osquery table).
+	Platform string `json:"platform"`
+}
+
+// getHostInfo retrieves system information about the host.
+//
+// On macOS and Linux it shells out to osqueryd to retrieve the information.
+//
+// On Windows:
+//
+//   - HardwareUUID is retrieved by shelling out to wmic, if that fails
+//     then the windows API are used.
+//   - HardwareSerial is currently not retrieved for Windows devices.
+//   - Hostname is retrieved using stdlib method.
+//   - Platform is always "windows" for windows hosts.
+//
+// NOTE: Windows uses a different approach to retrieve the device information
+// as there were issues at the time with shelling out to osquery - from what the
+// team remembers it would sometimes fail due to the osquery process not being ready yet.
+// A recent CI run without the Windows special-case did succeed, but since we don't
+// need the serial number for Windows at the moment, we opted to keep the
+// code as it is.
+func getHostInfo(osqueryPath string, osqueryDBPath string) (fleet.OrbitHostInfo, error) {
 	if runtime.GOOS == "windows" {
-		// NOTE: Windows uses a different approach as there were issues at the time
-		// with shelling out to osquery - from what the team remembers it would
-		// sometimes fail due to the osquery process not being ready yet. A recent
-		// CI run without the Windows special-case did succeed, but since we don't
-		// need the serial number for Windows at the moment, we opted to keep the
-		// code as it is.
 		uuidData, uuidSource, err := platform.GetSMBiosUUID()
 		if err != nil {
-			return "", "", err
+			return fleet.OrbitHostInfo{}, err
 		}
-
-		log.Debug().Msgf("UUID source was %s.", uuidSource)
-		return uuidData, "", nil
+		log.Debug().Str("source", string(uuidSource)).Msg("UUID")
+		// Hostname might differ from the one provided by osquery but we are sending it
+		// for troubleshooting purposes and to avoid empty host entries in the UI.
+		hostname, err := os.Hostname()
+		if err != nil {
+			return fleet.OrbitHostInfo{}, err
```

Comment:
Here you already have `uuidData, uuidSource`.
Wouldn't it be better to return them instead of empty `OrbitHostInfo ` ?

Reply:
Usually, convention is to return the zero value (`fleet.OrbitHostInfo{}`) if there's an error.
Otherwise it's hard to know what's set and what's not set in the returned struct (depending on what failed).

From the caller side:
```go
hostInfo, err := getHostInfo(...)
if err != nil {
  // ignore hostInfo and return or handle the error
}
// use the returned hostInfo (assured to be fully valid)
```

Reply:
Agree.